[BUG] Run `sort_values` on GPU encounter IndexError

Code:

```
raw = pd.DataFrame(np.random.rand(100, 10), columns=['a' + str(i) for i in range(10)])
mdf = DataFrame(raw, chunk_size=30).to_gpu()
mdf.sort_values(by='a0').execute()
```

Error stack:

```
mars/executor.py:720: in execute_tileable
    ret = self.execute_graph(chunk_graph, result_keys, n_parallel=n_parallel or n_thread,
mars/tests/core.py:678: in execute_graph
    return super().execute_graph(graph, keys, **kw)
mars/executor.py:693: in execute_graph
    res = graph_execution.execute(retval)
mars/executor.py:574: in execute
    future.result()
/opt/conda/lib/python3.8/concurrent/futures/_base.py:432: in result
    return self.__get_result()
/opt/conda/lib/python3.8/concurrent/futures/_base.py:388: in __get_result
    raise self._exception
/opt/conda/lib/python3.8/concurrent/futures/thread.py:57: in run
    result = self.fn(*self.args, **self.kwargs)
mars/tests/core.py:594: in _execute_operand
    super()._execute_operand(op)
mars/utils.py:451: in _inner
    return func(*args, **kwargs)
mars/executor.py:446: in _execute_operand
    self.handle_op(first_op, results, self._mock)
mars/executor.py:378: in handle_op
    return Executor.handle(*args, **kw)
mars/executor.py:644: in handle
    return runner(results, op)
mars/dataframe/sort/psrs.py:360: in execute
    ctx[op.outputs[-1].key] = res[by].iloc[slc]
/opt/conda/lib/python3.8/site-packages/cudf/core/indexing.py:179: in __getitem__
    return self._getitem_tuple_arg(arg)
/opt/conda/lib/python3.8/contextlib.py:75: in inner
    return func(*args, **kwds)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ __

self = <cudf.core.indexing._DataFrameIlocIndexer object at 0x7f6f55f93640>, arg = (slice(0, 24, 6),)

    @annotate("ILOC_GETITEM", color="blue", domain="cudf_python")
    def _getitem_tuple_arg(self, arg):
        from cudf import MultiIndex
        from cudf.core.column import column
        from cudf.core.index import as_index

        # Iloc Step 1:
        # Gather the columns specified by the second tuple arg
>       columns_df = self._get_column_selection(arg[1])
E       IndexError: tuple index out of range

/opt/conda/lib/python3.8/site-packages/cudf/core/indexing.py:393: IndexError
```

--- a/mars/dataframe/sort/core.py
+++ b/mars/dataframe/sort/core.py
@@ -27,14 +27,12 @@ class DataFrameSortOperand(DataFrameOperand):
     _parallel_kind = StringField('parallel_kind')
     _psrs_kinds = ListField('psrs_kinds', ValueType.string)
 
-    def __init__(self, axis=None, ascending=None, inplace=None, kind=None,
-                 na_position=None, ignore_index=None, parallel_kind=None, psrs_kinds=None, **kw):
-        super(DataFrameSortOperand, self).__init__(_axis=axis, _ascending=ascending,
-                                                   _inplace=inplace, _kind=kind,
-                                                   _na_position=na_position,
-                                                   _ignore_index=ignore_index,
-                                                   _parallel_kind=parallel_kind,
-                                                   _psrs_kinds=psrs_kinds, **kw)
+    def __init__(self, axis=None, ascending=None, inplace=None, kind=None, na_position=None,
+                 ignore_index=None, parallel_kind=None, psrs_kinds=None, gpu=False, **kw):
+        super().__init__(_axis=axis, _ascending=ascending, _inplace=inplace, _kind=kind,
+                         _na_position=na_position, _ignore_index=ignore_index,
+                         _parallel_kind=parallel_kind, _psrs_kinds=psrs_kinds,
+                         _gpu=gpu, **kw)
 
     @property
     def axis(self):

--- a/mars/dataframe/sort/psrs.py
+++ b/mars/dataframe/sort/psrs.py
@@ -39,10 +39,11 @@ class DataFramePSRSOperandMixin(DataFrameOperandMixin, PSRSOperandMixin):
         from .sort_values import DataFrameSortValues
         if isinstance(op, DataFrameSortValues):
             properties = dict(sort_type='sort_values', axis=op.axis, by=op.by, ascending=op.ascending,
-                              inplace=op.inplace, na_position=op.na_position)
+                              inplace=op.inplace, na_position=op.na_position, gpu=op.is_gpu())
         else:
             properties = dict(sort_type='sort_index', axis=op.axis, level=op.level, ascending=op.ascending,
-                              inplace=op.inplace, na_position=op.na_position, sort_remaining=op.sort_remaining)
+                              inplace=op.inplace, na_position=op.na_position, sort_remaining=op.sort_remaining,
+                              gpu=op.is_gpu())
         return properties
 
     @classmethod
@@ -272,11 +273,12 @@ class DataFramePSRSChunkOperand(DataFrameOperand):
     _n_partition = Int32Field('n_partition')
 
     def __init__(self, sort_type=None, by=None, axis=None, ascending=None, inplace=None, kind=None,
-                 na_position=None, level=None, sort_remaining=None, n_partition=None, output_types=None, **kw):
+                 na_position=None, level=None, sort_remaining=None, n_partition=None, gpu=None,
+                 output_types=None, **kw):
         super().__init__(_sort_type=sort_type, _by=by, _axis=axis, _ascending=ascending,
                          _inplace=inplace, _kind=kind, _na_position=na_position,
                          _level=level, _sort_remaining=sort_remaining, _n_partition=n_partition,
-                         _output_types=output_types, **kw)
+                         _gpu=gpu, _output_types=output_types, **kw)
 
     @property
     def sort_type(self):
@@ -329,6 +331,7 @@ class DataFramePSRSSortRegularSample(DataFramePSRSChunkOperand, DataFrameOperand
     @classmethod
     def execute(cls, ctx, op):
         a = ctx[op.inputs[0].key]
+        xdf = pd if isinstance(a, (pd.DataFrame, pd.Series)) else cudf
 
         if op.sort_type == 'sort_values':
             ctx[op.outputs[0].key] = res = execute_sort_values(a, op)
@@ -338,7 +341,7 @@ class DataFramePSRSSortRegularSample(DataFramePSRSChunkOperand, DataFrameOperand
         by = op.by
         add_distinct_col = bool(int(os.environ.get('PSRS_DISTINCT_COL', '0'))) \
             or getattr(ctx, 'running_mode', None) == RunningMode.distributed
-        if add_distinct_col and isinstance(a, pd.DataFrame) and op.sort_type == 'sort_values':
+        if add_distinct_col and isinstance(a, xdf.DataFrame) and op.sort_type == 'sort_values':
             # when running under distributed mode, we introduce an extra column
             # to make sure pivots are distinct
             chunk_idx = op.inputs[0].index[0]
@@ -350,10 +353,12 @@ class DataFramePSRSSortRegularSample(DataFramePSRSChunkOperand, DataFrameOperand
         n = op.n_partition
         if a.shape[op.axis] < n:
             num = n // a.shape[op.axis] + 1
-            res = execute_sort_values(pd.concat([res] * num), op, by=by)
+            res = execute_sort_values(xdf.concat([res] * num), op, by=by)
         w = int(res.shape[op.axis] // n)
 
-        slc = (slice(None),) * op.axis + (slice(0, n * w, w),)
+        slc = slice(0, n * w, w)
+        if op.axis == 1:
+            slc = (slice(None), slc)
         if op.sort_type == 'sort_values':
             # do regular sample
             if op.by is not None:
@@ -381,8 +386,9 @@ class DataFramePSRSConcatPivot(DataFramePSRSChunkOperand, DataFrameOperandMixin)
         p = len(inputs)
         assert a.shape[op.axis] == p ** 2
 
-        select = slice(p - 1, (p - 1) ** 2 + 1, p - 1)
-        slc = (slice(None),) * op.axis + (select,)
+        slc = slice(p - 1, (p - 1) ** 2 + 1, p - 1)
+        if op.axis == 1:
+            slc = (slice(None), slc)
         if op.sort_type == 'sort_values':
             a = execute_sort_values(a, op, inplace=False)
             ctx[op.outputs[-1].key] = a.iloc[slc]
@@ -468,43 +474,26 @@ class DataFramePSRSShuffle(DataFrameMapReduceOperand, DataFrameOperandMixin):
         a, pivots = [ctx[c.key] for c in op.inputs]
         out = op.outputs[0]
 
-        if isinstance(a, pd.DataFrame):
-            # use numpy.searchsorted to find split positions.
-            by = op.by
+        # use numpy.searchsorted to find split positions.
+        by = op.by
 
-            distinct_col = _PSRS_DISTINCT_COL if a.columns.nlevels == 1 \
-                else (_PSRS_DISTINCT_COL,) + ('',) * (a.columns.nlevels - 1)
-            if distinct_col in a.columns:
-                by += [distinct_col]
+        distinct_col = _PSRS_DISTINCT_COL if a.columns.nlevels == 1 \
+            else (_PSRS_DISTINCT_COL,) + ('',) * (a.columns.nlevels - 1)
+        if distinct_col in a.columns:
+            by += [distinct_col]
 
-            records = a[by].to_records(index=False)
-            p_records = pivots.to_records(index=False)
-            if op.ascending:
-                poses = records.searchsorted(p_records, side='right')
-            else:
-                poses = len(records) - records[::-1].searchsorted(p_records, side='right')
-            del records, p_records
+        records = a[by].to_records(index=False)
+        p_records = pivots.to_records(index=False)
+        if op.ascending:
+            poses = records.searchsorted(p_records, side='right')
+        else:
+            poses = len(records) - records[::-1].searchsorted(p_records, side='right')
+        del records, p_records
 
-            poses = (None,) + tuple(poses) + (None,)
-            for i in range(op.n_partition):
-                values = a.iloc[poses[i]: poses[i + 1]]
-                ctx[(out.key, str(i))] = values
-        else:  # pragma: no cover
-            # for cudf, find split positions in loops.
-            if op.ascending:
-                pivots.append(a.iloc[-1][op.by])
-                for i in range(op.n_partition):
-                    selected = a
-                    for label in op.by:
-                        selected = selected.loc[a[label] <= pivots.iloc[i][label]]
-                    ctx[(out.key, str(i))] = selected
-            else:
-                pivots.append(a.iloc[-1][op.by])
-                for i in range(op.n_partition):
-                    selected = a
-                    for label in op.by:
-                        selected = selected.loc[a[label] >= pivots.iloc[i][label]]
-                    ctx[(out.key, str(i))] = selected
+        poses = (None,) + tuple(poses) + (None,)
+        for i in range(op.n_partition):
+            values = a.iloc[poses[i]: poses[i + 1]]
+            ctx[(out.key, str(i))] = values
 
     @classmethod
     def _execute_series_map(cls, ctx, op):
@@ -554,7 +543,7 @@ class DataFramePSRSShuffle(DataFrameMapReduceOperand, DataFrameOperandMixin):
         raw_inputs = [ctx[(input_key, op.shuffle_key)] for input_key in input_keys]
         xdf = pd if isinstance(raw_inputs[0], (pd.DataFrame, pd.Series)) else cudf
         concat_values = xdf.concat(raw_inputs, axis=op.axis)
-        if isinstance(concat_values, pd.DataFrame):
+        if isinstance(concat_values, xdf.DataFrame):
             concat_values.drop(_PSRS_DISTINCT_COL, axis=1, inplace=True, errors='ignore')
 
         del raw_inputs[:]

--- a/mars/dataframe/sort/sort_index.py
+++ b/mars/dataframe/sort/sort_index.py
@@ -183,7 +183,7 @@ def sort_index(a, axis=0, level=None, ascending=True, inplace=False, kind='quick
     op = DataFrameSortIndex(level=level, axis=axis, ascending=ascending, inplace=inplace,
                             kind=kind, na_position=na_position, sort_remaining=sort_remaining,
                             ignore_index=ignore_index, parallel_kind=parallel_kind,
-                            psrs_kinds=psrs_kinds)
+                            psrs_kinds=psrs_kinds, gpu=a.op.is_gpu())
     sorted_a = op(a)
     if inplace:
         a.data = sorted_a.data

--- a/mars/dataframe/sort/sort_values.py
+++ b/mars/dataframe/sort/sort_values.py
@@ -217,7 +217,7 @@ def dataframe_sort_values(df, by, axis=0, ascending=True, inplace=False, kind='q
     by = by if isinstance(by, (list, tuple)) else [by]
     op = DataFrameSortValues(by=by, axis=axis, ascending=ascending, inplace=inplace, kind=kind,
                              na_position=na_position, ignore_index=ignore_index, parallel_kind=parallel_kind,
-                             psrs_kinds=psrs_kinds, output_types=[OutputType.dataframe])
+                             psrs_kinds=psrs_kinds, gpu=df.op.is_gpu(), output_types=[OutputType.dataframe])
     sorted_df = op(df)
     if inplace:
         df.data = sorted_df.data
@@ -312,7 +312,7 @@ def series_sort_values(series, axis=0, ascending=True, inplace=False, kind='quic
     op = DataFrameSortValues(axis=axis, ascending=ascending, inplace=inplace, kind=kind,
                              na_position=na_position, ignore_index=ignore_index,
                              parallel_kind=parallel_kind, psrs_kinds=psrs_kinds,
-                             output_types=[OutputType.series])
+                             output_types=[OutputType.series], gpu=series.op.is_gpu())
     sorted_series = op(series)
     if inplace:
         series.data = sorted_series.data

--- a/mars/dataframe/sort/tests/test_sort_execute.py
+++ b/mars/dataframe/sort/tests/test_sort_execute.py
@@ -19,7 +19,7 @@ import unittest
 import numpy as np
 import pandas as pd
 
-from mars.tests.core import ExecutorForTest
+from mars.tests.core import ExecutorForTest, require_cudf
 from mars.dataframe import DataFrame, Series, ArrowStringDtype
 from mars.session import new_session
 
@@ -281,3 +281,42 @@ class Test(unittest.TestCase):
         result = self.executor.execute_dataframe(df, concat=True)[0]
         expected = raw.sort_values(by='b')
         pd.testing.assert_frame_equal(result, expected)
+
+    @require_cudf
+    def testGPUExecution(self):
+        # test sort_values
+        distinct_opts = ['0'] if sys.platform.lower().startswith('win') else ['0', '1']
+        for add_distinct in distinct_opts:
+            os.environ['PSRS_DISTINCT_COL'] = add_distinct
+
+            # test dataframe
+            raw = pd.DataFrame(np.random.rand(100, 10), columns=['a' + str(i) for i in range(10)])
+            mdf = DataFrame(raw, chunk_size=30).to_gpu()
+
+            result = self.executor.execute_dataframe(mdf.sort_values(by='a0'), concat=True)[0]
+            expected = raw.sort_values(by='a0')
+            pd.testing.assert_frame_equal(result.to_pandas(), expected)
+
+            # test series
+            raw = pd.Series(np.random.rand(10))
+            series = Series(raw).to_gpu()
+
+            result = self.executor.execute_dataframe(series.sort_values(), concat=True)[0]
+            expected = raw.sort_values()
+            pd.testing.assert_series_equal(result.to_pandas(), expected)
+
+        # test DataFrame.sort_index
+        raw = pd.DataFrame(np.random.rand(10, 10), columns=np.random.rand(10))
+        mdf = DataFrame(raw).to_gpu()
+
+        result = self.executor.execute_dataframe(mdf.sort_index(), concat=True)[0]
+        expected = raw.sort_index()
+        pd.testing.assert_frame_equal(result.to_pandas(), expected)
+
+        # test Series.sort_index
+        raw = pd.Series(np.random.rand(10, ), index=np.random.rand(10))
+        series = Series(raw).to_gpu()
+
+        result = self.executor.execute_dataframe(series.sort_index(), concat=True)[0]
+        expected = raw.sort_index()
+        pd.testing.assert_series_equal(result.to_pandas(), expected)

